### PR TITLE
api(cdp): newCDPSession accepts frames, too

### PR DESCRIPTION
**Pull request by rwoll**
Without this, Playwright's CDP feature leaves unreachable
targets (namely OOPIFs).

This change allows for more advanced experimentation in user-land
without relying on out-of-band CDP connections and clients.

Now you can, for example, call `DOM.getDocument` on the
page OR main frame, observe there is an iframe node with
no `contentDocument` (i.e. OOPIF), make note of the referenced
`frameId`, and then iterate of page.frames() calling `Target.getInfo`
on each to link the Playwright Frame with the CDP `frameId` and
then recurse.

Relates #8113

### CODE REVIEW

File: src/server/chromium/crBrowser.ts [hunk 1]
```diff
@@ -503,10 +504,18 @@ export class CRBrowserContext extends BrowserContext {
...
+    if (target instanceof Page) {
+      targetId = (target._delegate as CRPage)._targetId;
+    } else if (target instanceof Frame) {
+      const session = (target._page._delegate as CRPage)._sessions.get(target._id);
+      if (!session) throw new Error('No session for Target ID. This frame might be part of another target');
+      targetId = session._targetId;
+    } else {
+      throw new Error('target: expected Page or Frame');
+    }
+
```
Comment (rwoll) [inline src/server/chromium/crBrowser.ts:518]: All this could be changed to:
```
const targetId = (target instanceof Page ? target.mainFrame() : target)._id
```
if we don't mind getting `browserContext.newCDPSession: Protocol error (Target.attachToTarget): No target with given id found` for frames that are part of another target. (Since the original code didn't do this, I went the more explicit _targetId finding route).

File: src/protocol/protocol.yml [hunk 2]
```diff
@@ -629,7 +629,8 @@ BrowserContext:
 
     newCDPSession:
       parameters:
-        page: Page
+        page: Page?
+        frame: Frame?
```
Comment (rwoll) [inline src/protocol/protocol.yml:633]: Is there a union type/notation to match the TS?

**Comment by rwoll** [comment]
@pavelfeldman This is ready for a review pass! 🎉 Thanks!

**Comment by vigneshshanmugam** [comment]
@rwoll This API looks interesting, Curious if the proposed API would also address the issue I mentioned here - https://github.com/microsoft/playwright/issues/6340#issuecomment-829647616

**Comment by rwoll** [comment]
> @rwoll This API looks interesting, Curious if the proposed API would also address the issue I mentioned here - https://github.com/microsoft/playwright/issues/6340#issuecomment-829647616

@vigneshshanmugam Can you rephrase the question and explain the particular information you're still after? I skimmed through that issue thread, but wasn't quite sure which bits to answer about. Thanks!

**Comment by vigneshshanmugam** [comment]
> Can you rephrase the question and explain the particular information you're still after? I skimmed through that issue thread, but wasn't quite sure which bits to answer about. Thanks!

@rwoll Sure, Currently all user level created CDP sessions are tied to page which means when new frames are created we will miss out on the events that are requested on the new page. Lets take an example
```ts
const context = await browser.newContext(playwrightOptions);
const page = await context.newPage();
const client = await context.newCDPSession(page);

client.send('Network.enable'); 

client.on('Network.requestWillBeSent', () => {
    // This wont get logged for iframes or pages in new tabs
});
```

I was curios if the proposed API allows for monitoring these subsequent pages. Thanks.

File: src/server/chromium/crBrowser.ts [hunk 3]
```diff
@@ -503,10 +504,18 @@ export class CRBrowserContext extends BrowserContext {
...
-    if (!(page instanceof Page))
-      throw new Error('page: expected Page');
-    const targetId = (page._delegate as CRPage)._targetId;
+  async newCDPSession(target: Page | Frame): Promise<CRSession> {
+    let targetId: string | null = null;
+    if (target instanceof Page) {
+      targetId = (target._delegate as CRPage)._targetId;
+    } else if (target instanceof Frame) {
+      const session = (target._page._delegate as CRPage)._sessions.get(target._id);
+      if (!session) throw new Error('No session for Target ID. This frame might be part of another target');
```
Comment (dgozman) [inline src/server/chromium/crBrowser.ts:513]: `throw new Error('This frame does not have a separate CDP session, it is a part of the parent frame's session')`

**Comment by rwoll** [comment]
> > Can you rephrase the question and explain the particular information you're still after? I skimmed through that issue thread, but wasn't quite sure which bits to answer about. Thanks!
> 
> 
> 
> @rwoll Sure, Currently all user level created CDP sessions are tied to page which means when new frames are created we will miss out on the events that are requested on the new page. Lets take an example
> 
> ```ts
> 
> const context = await browser.newContext(playwrightOptions);
> 
> const page = await context.newPage();
> 
> const client = await context.newCDPSession(page);
> 
> 
> 
> client.send('Network.enable'); 
> 
> 
> 
> client.on('Network.requestWillBeSent', () => {
> 
>     // This wont get logged for iframes or pages in new tabs
> 
> });
> 
> ```
> 
> 
> 
> I was curios if the proposed API allows for monitoring these subsequent pages. Thanks. 

As is, this won't work, but you could possibly set up sessions on each of the new frames/pages. Still, even that I don't think will work perfectly without race/timing issues so you might miss requests. You can experiment with `Target.setAutoAttach`'s `waitForDebuggerOnStart` to allow yourself time to set up handlers on the new frames before you send `Runtime.runIfWaitingForDebugger`, but that may deadlock coupled with the PW code.

File: tests/chromium/session.spec.ts [hunk 4]
```diff
@@ -37,10 +37,10 @@ it('should send events', async function({page, server}) {
   expect(events.length).toBe(1);
 });
 
-it('should only accept a page', async function({page}) {
-  // @ts-expect-error newCDPSession expects a Page
+it('should only accept a page or frame', async function({page}) {
+  // @ts-expect-error newCDPSession expects a Page or Frame
   const error = await page.context().newCDPSession(page.context()).catch(e => e);
-  expect(error.message).toContain('page: expected Page');
+  expect(error.message).toContain('frame: expected Frame');
```
[the remark this answers is not in the document]
Comment (rwoll) [reply tests/chromium/session.spec.ts]: I "fixed" this up. The error was automagically coming from channelOwner.ts since context is channelable. Looks like the error might depend on how the object props are iterated!

**Comment by vigneshshanmugam** [comment]
> @vigneshshanmugam Can you file a separate request for this use case? (Or re-open your linked issue.) I'm happy to brainstorm further but want to separate that discussion from this PR. Thanks!

Absolutely, Apologies for opening up the discussion here, Let's chat in a separate issue. Thanks!

**Review by dgozman — APPROVED** [review]
